jupyter_accessibility/project.py

```diff
@@ -0,0 +1,428 @@
+                )
+
+    def pip_install(self):
+        """pip install the python package"""
+        yield dict(
+            basename="install",
+            name=f"pip:{self.name}",
+            actions=[
+                do(f"{self.conda} python -m pip install jupyter_packaging"),
+                do(
+                    f"{self.conda} python -m pip install -e.",  # many settings are held in the env
+                    cwd=self.path,
+                    env=PIP_BUILD_ENV,
+                ),
+            ],
+        )
+
+    def tasks(self):
+        yield self.clone()
+        # we need to clone before we can do anything
+        # these have to happen afterwards. the order in the method has no bearing on the execution order
+        yield self.yarn_install()
+
+    def get_repo(self, id):
+        if isinstance(id, type):
+            id = id.id
+
+        return self.parent.repos[id]
+
+
+# documentation tests and unit tests would have different task specifications
+class PlayWrightTests:
+    """a mixin with playwright tests"""
+
+    def tests(self):
+        target = self.dir / AXE_TEMPLATE.name
+        yield dict(
+            basename="test_setup",
+            name="copy-templates",
+            actions=[(mv, [AXE_TEMPLATE, target])],
+            clean=[(rmdir, [target])],
+            uptodate=[target.exists()],
+        )
+        yield dict(
+            basename="test_setup",
+            name="pip",
+            actions=[do(f"{self.conda} pip install -r requirements.txt", cwd=target)],
+        )
+        yield dict(
+            basename="test_setup",
+            name="yarn",
+            actions=[do(f"{self.conda} yarn install", cwd=target)],
+        )
+        if CI or GITPOD:
+            # i dont know how to measure the state of playwright browsers. this measure should go in the
+            # uptodate attribute of the task. currently, we're just using whatever logic. in theory, we could
+            # just run this and rely on whatever caching playwright does.
+            # why are there so many browsers?
+            yield dict(
+                basename="test_setup",
+                name="playwright",
+                actions=[
+                    do(f"{self.conda} npx playwright install --with-deps chromium", cwd=target)
+                ],
+            )
+        yield dict(
+            basename="test",
+            name="playwright",
+            actions=[
+                do(f"echo 'the target is {target.absolute()}'"),
+                do(f"{self.conda} npx playwright test", cwd=target),
+            ],
+        )
+
+
+class JupyterApplication(Common, PlayWrightTests):
+    pass
+
+
+class JupyterLabPackage(Package, JupyterApplication, id="jupyterlab"):
+    pass
+
+
+class RetroLabPackage(Package, JupyterApplication, id="retrolab"):
+    """a class for the retrolab package from pypi"""
+
+
+class JupyterLabRepo(Repo, JupyterApplication, id="jupyterlab"):
```

Comment:
do we need to have all these separate classes?
- or rather is there a way to define a class dynamically? food for thought for later iterations

Reply:
yea. i think we'll see if this structure survives. extensions are going to test its efficacy